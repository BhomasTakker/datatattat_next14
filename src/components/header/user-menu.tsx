"use client";

import { PATHS } from "@/lib/routing/paths";
import Link from "next/link";
import { BiSolidUser } from "react-icons/bi";
import styles from "./user-menu.module.scss";
import { useRef } from "react";
import { useSession } from "next-auth/react";
import { usePathname } from "next/navigation";

export const UserMenu = () => {
	const { data } = useSession();
	const pathname = usePathname();
	const modalRef = useRef<HTMLDialogElement>(null);

	const openModal = () => {
		if (modalRef.current) {
			modalRef.current.showModal();
		}
	};
	const closeModal = () => {
		if (modalRef.current) {
			modalRef.current.close();
		}
	};

	return (
		<div>
			<button className={styles.menuButton} onClick={openModal}>
				<BiSolidUser size={"2rem"} />
			</button>
			<dialog ref={modalRef} className={styles.menu} onBlur={closeModal}>
				<ul>
					{/* // callback to current page */}
					<li>
						<Link href={PATHS.signOut(pathname)}>Sign Out</Link>
					</li>
					<li>
						{/* add page query? */}
						<Link href={PATHS.edit()}>Edit</Link>
					</li>
					{data?.user?.name ? (
						<li>
							<Link href={PATHS.user(data?.user?.name || "")}>User Page</Link>
						</li>
					) : null}
				</ul>
			</dialog>
		</div>
	);
};
