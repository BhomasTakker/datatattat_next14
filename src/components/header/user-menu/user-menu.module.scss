@use "../../../scss/abstracts/colours/" as *;
@use "../../../scss/abstracts/units/" as *;
@use "../../../scss/abstracts/font/" as *;

.menuButton {
	border: none;
	background-color: unset;
	cursor: pointer;
	padding: $spacingXs;
	border-radius: 50%;

	.logo {
		font-size: $title-md;
		color: $textColourDark;
	}

	&:hover {
		background-color: $shadowLight1;
	}
}
