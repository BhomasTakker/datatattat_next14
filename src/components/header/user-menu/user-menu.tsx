"use client";

// This could be a generic component eaasily enough
import { BiSolidUser } from "react-icons/bi";
import styles from "./user-menu.module.scss";
import React, { useEffect, useState } from "react";
import { DropDownProps, MenuDropDown } from "./drop-down/menu-drop-down";

type UserMenuProps = {
	username: string;
	avatar: string;
};

export const UserMenu = ({ username, avatar }: UserMenuProps) => {
	const [isOpen, setIsOpen] = useState(false);

	useEffect(() => {
		const closeHnd = () => setIsOpen(false);
		if (isOpen) {
			document.addEventListener("click", closeHnd);
		}
		return () => {
			document.removeEventListener("click", closeHnd);
		};
	}, [isOpen]);

	return (
		<div data-testid="user-menu">
			<button className={styles.menuButton} onClick={() => setIsOpen(!isOpen)}>
				<BiSolidUser className={styles.logo} />
			</button>
			<MenuDropDown
				isOpen={isOpen}
				setIsOpen={setIsOpen}
				username={username}
				avatar={avatar}
			/>
		</div>
	);
};
