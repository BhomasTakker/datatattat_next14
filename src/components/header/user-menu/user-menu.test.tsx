import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { UserMenu } from "./user-menu";

const menuProps = {
	username: "John Cena",
	avatar: "https://example.com/avatar.jpg",
};

// We should probaly mock this guy somewhere
jest.mock("next-auth/react", () => {
	return {
		useSession: jest.fn().mockResolvedValue({
			data: {
				user: {
					name: "John Cena",
				},
			},
		}),
	};
});

jest.mock("next/navigation", () => {
	return {
		usePathname: jest.fn().mockResolvedValue("/mock-route"),
	};
});

// problem with useRef - showModal does not exist.....
// mocking/spying no work....
describe("User Menu", () => {
	it("renders a button to open the user menu", () => {
		render(
			<UserMenu username={menuProps.username} avatar={menuProps.avatar} />
		);
		const button = screen.getByRole("button");
		expect(button).toBeInTheDocument();
	});
	// it("renders a user menu with a sign out link", () => {
	// 	render(<UserMenu />);
	// 	const button = screen.getByRole("button");
	// 	button.click();
	// 	const link = screen.queryByText(/sign out/i);
	// 	expect(link).toBeInTheDocument();
	// });
	it("renders unopened user menu unchanged", () => {
		const { container } = render(
			<UserMenu username={menuProps.username} avatar={menuProps.avatar} />
		);
		expect(container).toMatchSnapshot();
	});

	// it("renders opened user menu unchanged", () => {
	// 	const { container } = render(<UserMenu />);
	// 	const button = screen.getByRole("button");
	// 	button.click();
	// 	expect(container).toMatchSnapshot();
	// });
});
