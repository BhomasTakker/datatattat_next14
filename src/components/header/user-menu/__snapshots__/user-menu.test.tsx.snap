// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Menu renders unopened user menu unchanged 1`] = `
<div>
  <div>
    <button
      class="menuButton"
    >
      <svg
        class="logo"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.5 6.5C7.5 8.981 9.519 11 12 11s4.5-2.019 4.5-4.5S14.481 2 12 2 7.5 4.019 7.5 6.5zM20 21h1v-1c0-3.859-3.141-7-7-7h-4c-3.86 0-7 3.141-7 7v1h17z"
        />
      </svg>
    </button>
    <dialog
      class="menu"
    >
      <ul>
        <li>
          <a
            href="/api/auth/signout?callbackUrl=[object Promise]"
          >
            Sign Out
          </a>
        </li>
        <li>
          <a
            href="/edit"
          >
            Edit
          </a>
        </li>
      </ul>
    </dialog>
  </div>
</div>
`;
